drivers/include/periph/usbdev.h

```diff
@@ -0,0 +1,559 @@
+ * approach where multiple USB functions/interfaces can be multiplexed over a
+ * single USB interface. Each interface can be implemented as a separate module.
+ * The interface handler does not have to care about the usb device itself or
+ * it's endpoint number. It can simply request an available endpoint from the
+ * usb device with the @ref usbdev_new_ep function.
+ *
+ * Data transmission is done by requesting the endpoint with a max packet size.
+ * Memory for this buffer is allocated from dedicated memory of the MCU USB
+ * peripheral or from a buffer allocated by the peripheral specific usbdev
+ * struct. Received data from the host ends up at this buffer automatically
+ * by the low level drivers. Signalling that the data at the specified address
+ * is ready to be reused is done with the @ref usbdev_ep_ready function by
+ * supplying a size of 0 for the @p len argument.
+ *
+ * For transmitting data back to the host, a similar approach is used. The data
+ * to be transmitted is written to the specified address and the
+ * @ref usbdev_ep_ready function is called with the size of the data as @p len
+ * argument.
+ *
+ * This approach of setting the address and only indicating new data available
+ * is done to allow the low level USB peripheral to use DMA to transfer the data
+ * from and to the MCU memory.
+ *
+ * A callback function is required for signalling events from the driver. The
+ * @ref USBDEV_EVENT_ESR is special in that it indicates that the USB peripheral
+ * had an interrupt that needs to be serviced in a non-interrupt context. This
+ * requires the usbdev_esr function to be called. In the case that the event is
+ * signalled via the @ref usbdev_t::epcb callback, the @ref usbdev_ep_esr
+ * should be called to service the endpoint specific events from the
+ * peripheral.
+ *
+ * @file
+ * @brief       Definitions low-level USB driver interface
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ */
+
+#ifndef PERIPH_USBDEV_H
+#define PERIPH_USBDEV_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "assert.h"
+#include "usb.h"
+#include "usb/usbopt.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief usbdev_t forward declaration
+ */
+typedef struct usbdev usbdev_t;
+
+/**
+ * @brief usbdev_ep_t forward declaration
+ */
+typedef struct usbdev_ep usbdev_ep_t;
+
+/**
+ * @brief Statically allocated buffer space for endpoints.
+ *
+ * When the device doesn't have dedicated memory for endpoint buffers, a
+ * buffer of this size is allocated to contain the endpoint buffers. Only
+ * needs to be as big as the total buffer space required by all endpoints
+ */
+#ifndef USBDEV_EP_BUF_SPACE
+#define USBDEV_EP_BUF_SPACE     1024
+#endif
+
+/**
+ * @brief Number of USB IN and OUT endpoints allocated
+ *
+ * Configures the number of endpoints allocated. An equal number of IN and OUT
+ * endpoints are allocated
+ */
+#ifndef USBDEV_NUM_ENDPOINTS
+#define USBDEV_NUM_ENDPOINTS       8
```

Comment:
What about define it within each cpu/ ? This will force user to ensure the MCU can handle them. We might have surprise in the future if we keep a default value here.

Reply:
The idea here was to have a global value to set the number of IN and OUT endpoints. This way, an application that only needs a limited number of endpoints can set the required number in a platform-independent way.

I agree that this might cause issues. We could move this to something like `cpu.h`, but I can't say I really like that solution. I don't mind keeping it this way at the moment and solve the issue in a follow up as soon as we have a few drivers.